src/IR/LambdaLift.hs

```diff
@@ -0,0 +1,205 @@
+{-# LANGUAGE DerivingVia #-}
+module IR.LambdaLift where
+
+import qualified Common.Compiler               as Compiler
+import           Common.Identifiers
+import qualified IR.IR                         as I
+
+import qualified IR.Types.Poly                 as Poly
+
+import           Control.Monad.Except           ( MonadError(..) )
+import           Control.Monad.State.Lazy       ( MonadState
+                                                , StateT(..)
+                                                , evalStateT
+                                                , gets
+                                                , modify
+                                                , unless
+                                                )
+
+import qualified Data.Bifunctor                as B
+import qualified Data.Foldable                 as F
+import qualified Data.Map                      as M
+import           Data.Maybe                     ( catMaybes )
+import qualified Data.Set                      as S
+
+data LiftCtx = LiftCtx
+  { globalScope  :: S.Set I.VarId
+  , currentScope :: S.Set I.VarId
+  , freeTypes    :: M.Map I.VarId Poly.Type
+  , lifted       :: [(I.VarId, I.Expr Poly.Type)]
+  , anonCount    :: Int
+  }
+
+newtype LiftFn a = LiftFn (StateT LiftCtx Compiler.Pass a)
+  deriving Functor                      via (StateT LiftCtx Compiler.Pass)
+  deriving Applicative                  via (StateT LiftCtx Compiler.Pass)
+  deriving Monad                        via (StateT LiftCtx Compiler.Pass)
+  deriving MonadFail                    via (StateT LiftCtx Compiler.Pass)
+  deriving (MonadError Compiler.Error)  via (StateT LiftCtx Compiler.Pass)
+  deriving (MonadState LiftCtx)         via (StateT LiftCtx Compiler.Pass)
+
+exprType :: I.Expr Poly.Type -> Poly.Type
+exprType (I.Lambda _ _ t) = t
+exprType (I.App    _ _ t) = t
+exprType (I.Lit _ t     ) = t
+exprType (I.Var _ t     ) = t
+exprType (I.Prim _ _ t  ) = t
+exprType (I.Let  _ _ t  ) = t
+exprType (I.Data _ t    ) = t
+exprType (I.Match _ _ t ) = t
+
+zipArgsWithArrow :: [Binder] -> Poly.Type -> [(Binder, Poly.Type)]
```

Comment:
Can't this be rewritten as:

```haskell
zip bs $ fst collectArrow
```

I recall I had written a `collectArrow :: Type -> ([Type], Type)` that unpacks an arrow type into its arguments and its return type (returns empty list for non-arrow type).

Also try to avoid throwing `error` here. If you need, throw a compiler error using utilities from `Compiler.Pass`.